--- a/src/rewind/rewind.c
+++ b/src/rewind/rewind.c
@@ -30,11 +30,9 @@
 #include "postmaster/bgwriter.h"
 #include "postmaster/autovacuum.h"
 #include "storage/bufmgr.h"
-#include "storage/ipc.h"
 #include "storage/latch.h"
 #include "storage/proc.h"
 #include "storage/sinvaladt.h"
-#include "utils/elog.h"
 #include "utils/memutils.h"
 #include "utils/snapmgr.h"
 #include "utils/syscache.h"
@@ -81,9 +79,6 @@
 #define	REWIND_MODE_XID	(3)
 
 static void orioledb_rewind_internal(int rewind_mode, int rewind_time, OXid rewind_oxid, TransactionId rewind_xid, TimestampTz rewind_timestamp);
-static void try_restart_pg(void);
-static void cleanup_fds(void);
-static void bootstrap_signals(void);
 
 /* Interface functions */
 
@@ -790,8 +785,9 @@
 
 	LWLockRelease(&rewindMeta->rewindEvictLock);
 	elog(LOG, "Rewind complete");
-
-	try_restart_pg();
+	/* Shutdown Postgres */
+	(void) kill(PostmasterPid, SIGTERM);
+	return;
 }
 
 TransactionId
@@ -1693,144 +1689,3 @@
 {
 	return pg_atomic_read_u64(&rewindMeta->runXmin);
 }
-
-
-/*
- * try_restart_pg
- *
- * Attempt to restart the PostgreSQL instance.
- *
- * This function spawns a persistent, detached process that executes
- * "pg_ctl restart". Detachment ensures the child process survives
- * once the postmaster exits.
- *
- * dependency: This implementation relies on fork(2) and setsid(2). Systems
- * lacking these primitives (e.g. Windows), are not supported, so we can't end
- * up here. See orioledb_enable_rewind_check_hook for details.
- */
-static void
-try_restart_pg(void)
-{
-	pid_t		pid;
-	sigset_t	blocked_sigset,
-				old_sigset;
-
-	/* Block all signals until we're done */
-	sigfillset(&blocked_sigset);
-	sigprocmask(SIG_BLOCK, &blocked_sigset, &old_sigset);
-
-	/* Flush stdio channels just before fork, to avoid double-output problems */
-	fflush(NULL);
-
-	pid = fork();
-
-	if (pid < 0)
-	{
-		/* fork failed, restore signals and bail */
-		sigprocmask(SIG_SETMASK, &old_sigset, NULL);
-
-		elog(DEBUG3, "fork failed while attempting to restart,"
-			 " stopping instead");
-
-		ereport(WARNING,
-				(errmsg("could not restart instance"),
-				 errdetail("Sending shutdown request to postmaster.")));
-		(void) kill(PostmasterPid, SIGTERM);
-		return;
-	}
-
-	if (pid > 0)
-	{
-		/* fork successful, in parent. Restore signals and return to client */
-		sigprocmask(SIG_SETMASK, &old_sigset, NULL);
-		ereport(NOTICE,
-				(errmsg("attempting to restart database system")));
-		return;
-	}
-
-	/* fork succeeded, in child */
-	if (setsid() < 0)
-	{
-		/* setsid failed: */
-		goto emergency_shutdown;
-	}
-
-	{
-		/* We're in a session that will survive when the parent goes away */
-		sigset_t	empty_mask;
-		char		bindir[MAXPGPATH];
-		char		cmd[PG_CTL_MAX_CMD_LEN];
-		char	   *lastslash;
-
-		strlcpy(bindir, my_exec_path, MAXPGPATH);
-
-		lastslash = strrchr(bindir, '/');
-
-		/* if for some reason we got a bogus bindir */
-		if (lastslash == NULL)
-			goto emergency_shutdown;
-
-		*lastslash = '\0';
-		snprintf(cmd, sizeof(cmd), "%s/pg_ctl", bindir);
-
-		/* Do a little dance with fds to make logger shutdown cleanly */
-		cleanup_fds();
-		/* Sleep 0.5s to let the parent return */
-		pg_usleep(500000);
-		/* Be paranoid: restore default signal handlers and mask before execl */
-		bootstrap_signals();
-		sigemptyset(&empty_mask);
-		sigprocmask(SIG_SETMASK, &empty_mask, NULL);
-
-		execl(cmd, cmd, "restart", "-D", DataDir, (char *) NULL);
-		/* execl failed: */
-		goto emergency_shutdown;
-	}
-
-emergency_shutdown:
-
-	/*
-	 * If we got here, either execl or setsid failed. We can't just bail
-	 * because logging isn't safe since we can't guarantee allocating memory
-	 * won't result in a deadlock. Additionally we might've already cleaned up
-	 * the FDs so elog won't reach the logger anyway. We also can't really
-	 * proc_exit() or even exit(): Both will result in a crash while executing
-	 * on-exit callbacks. So we request a shutdown and _exit(). 71 exit code
-	 * is "system error". We don't want to put too much effort into
-	 * investigating here.
-	 */
-	(void) kill(PostmasterPid, SIGTERM);
-	_exit(71);
-}
-
-static void
-cleanup_fds(void)
-{
-	int			devnull = open("/dev/null", O_RDWR);
-
-	if (devnull >= 0)
-	{
-		/*
-		 * We can't just close stderr/stdout fds, so redirect them to
-		 * /dev/null instead
-		 */
-		dup2(devnull, fileno(stdin));
-		dup2(devnull, fileno(stderr));
-		dup2(devnull, fileno(stdout));
-
-		/* Be paranoid: we don't want to ever close stdin/stderr/stdout */
-		if (devnull > fileno(stdin) && devnull > fileno(stdout)
-			&& devnull > fileno(stderr))
-			close(devnull);
-	}
-}
-
-static void
-bootstrap_signals(void)
-{
-	pqsignal(SIGHUP, SIG_DFL);
-	pqsignal(SIGPIPE, SIG_DFL);
-	pqsignal(SIGINT, SIG_DFL);
-	pqsignal(SIGTERM, SIG_DFL);
-	pqsignal(SIGQUIT, SIG_DFL);
-}
--- a/include/rewind/rewind.h
+++ b/include/rewind/rewind.h
@@ -42,10 +42,6 @@
 
 #define SUBXIDS_PER_ITEM	(25)
 
-
-#define PG_CTL_CMD_LEN (8)		/* Actually we only need 4 extra chars */
-#define PG_CTL_MAX_CMD_LEN (MAXPGPATH + PG_CTL_CMD_LEN)
-
 /* RewindItem and SubxidsItem should have same size to be castable to each other */
 /* Empty RewindItem and SubxidsItem have invalid oxid and tag */
 typedef struct RewindItem
--- a/src/orioledb.c
+++ b/src/orioledb.c
@@ -269,32 +269,6 @@
 	.rm_mask = NULL,
 	.rm_decode = orioledb_decode
 };
-
-/*
- * We currently do not support restarting PG instance from within the extension
- * on certain systems. Refuse to enable rewind on those systems.
- */
-static bool
-orioledb_enable_rewind_check_hook(bool *newval, void **extra, GucSource source)
-{
-#if defined(WIN32)
-	if (*newval)
-	{
-		GUC_check_errcode(ERRCODE_FEATURE_NOT_SUPPORTED);
-		GUC_check_errdetail("Rewind is not supported on Windows.");
-		return false;
-	}
-#elif !defined(HAVE_SETSID)
-	if (*newval)
-	{
-		GUC_check_errcode(ERRCODE_FEATURE_NOT_SUPPORTED);
-		GUC_check_errdetail("Rewind is not supported on systems without setsid(2).");
-		return false;
-	}
-#endif
-	/* Supported system or newval == false */
-	return true;
-}
 
 void
 _PG_init(void)
@@ -853,7 +827,7 @@
 							 false,
 							 PGC_POSTMASTER,
 							 0,
-							 orioledb_enable_rewind_check_hook,
+							 NULL,
 							 NULL,
 							 NULL);
 
