diff --git a/src/constrained_extensions.c b/src/constrained_extensions.c
index fc1b270..09697e8 100644
--- a/src/constrained_extensions.c
+++ b/src/constrained_extensions.c
@@ -163,7 +163,7 @@ parse_constrained_extensions(
 
        json_constrained_extension_parse_state state = {JCE_EXPECT_TOPLEVEL_START, NULL, 0, cexts};
 
-       lex = makeJsonLexContextCstringLen(pstrdup(str), strlen(str), PG_UTF8, true);
+       lex = makeJsonLexContextCstringLen(NULL, pstrdup(str), strlen(str), PG_UTF8, true);
 
        sem.semstate = &state;
        sem.object_start = json_object_start;
@@ -181,6 +181,8 @@ parse_constrained_extensions(
        if (json_error != JSON_SUCCESS)
                state.error_msg = "invalid json";
 
+       pfree(lex);
+
        return state;
 }